src/scalaDebugger.ts

```diff
@@ -46,15 +50,169 @@ export async function start(
 
 class ScalaConfigProvider implements vscode.DebugConfigurationProvider {
   provideDebugConfigurations(): ProviderResult<DebugConfiguration[]> {
-    return [];
+    const mainClassPick = "Main Class";
+    const testClassPick = "Test Suite";
+
+    return vscode.window
+      .showQuickPick([mainClassPick, testClassPick], {
+        placeHolder: "Pick kind of class to debug",
+      })
+      .then((result) => {
+        if (result === mainClassPick) {
+          return this.provideDebugMainClassConfiguration().then((config) => {
+            if (config === undefined) return [];
+            return [config];
+          });
+        } else if (result === testClassPick) {
+          return this.provideDebugTestClassConfiguration().then((config) => {
+            if (config === undefined) return [];
+            return [config];
+          });
+        }
+        return [];
+      });
   }
 
   resolveDebugConfiguration(
     _folder: WorkspaceFolder | undefined,
     debugConfiguration: DebugConfiguration
   ): ProviderResult<DebugConfiguration> {
+    if (debugConfiguration.type === undefined) return null;
     return debugConfiguration;
   }
+
+  private provideDebugMainClassConfiguration(): Thenable<
+    DebugConfiguration | undefined
+  > {
+    return this.askForBuildTarget().then((buildTarget) => {
+      if (buildTarget === undefined) return undefined;
+      return this.askForClassName().then((className) => {
+        if (className === undefined) return undefined;
+        return this.askForArguments().then((args) => {
+          if (args === undefined) return undefined;
+          return this.askForConfigurationName(className).then((name) => {
+            if (name === undefined) return undefined;
+            const result: DebugConfiguration = {
+              type: configurationType,
+              name: name,
+              request: launchRequestType,
+              mainClass: className,
+              buildTarget: buildTarget,
+              args: args,
+            };
+            return result;
+          });
+        });
+      });
+    });
+  }
```

Comment:
suggestion: instead of using `undefined` to represent the exceptional case, maybe we can use `Promise.reject()` and then catch the `Promise` above (see my other comment).

This would avoid checking for `undefined` at every step

Reply:
This would become

```suggestion
  private provideDebugMainClassConfiguration(): Thenable<DebugConfiguration> {
    return this.askForBuildTarget().then((buildTarget) =>
      this.askForClassName().then((className) =>
        this.askForArguments().then((args) =>
          this.askForConfigurationName(className).then((name) =>
            const result: DebugConfiguration = {
              type: configurationType,
              name: name,
              request: launchRequestType,
              mainClass: className,
              buildTarget: buildTarget,
              args: args,
            };
            return result;
          );
        );
      );
    );
  }
```

(provided that all sub-functions return `Thenable<A>` instead of `Thenable<A | undefined>` and fail with `Promise.reject` instead)

Reply:
done.